File: leetcode/BFS/leetcode-0127/127.java
import java.util.HashSet;

class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        HashSet<String> set = new HashSet<>(wordList);
        if (!set.contains(endWord)) {
            return 0;
        }

        Map<String, Integer> map = new HashMap<>();
        map.put(beginWord, 1);
        Queue<String> q = new LinkedList<>();
        q.offer(beginWord);

        while (!q.isEmpty()) {
            String word = q.poll();
            for (int i = 0; i < word.length(); i++) {
                char[] ch = word.toCharArray();
                for (char c = 'a'; c <= 'z'; c++) {
                    ch[i] = c;
                    String newWord = new String(ch);
                    if (set.contains(newWord) && newWord.equals(endWord)) {
                        return map.get(word) + 1;
                    }

                    if (set.contains(newWord) && !map.containsKey(newWord)) {
                        map.put(newWord, map.get(word) + 1);
                        q.poll(newWord);
                    }
                }
            }
        }
        return 0;
    }
}

File: leetcode/BFS/leetcode-0127/127.go
package bfs

func ladderLength(beginWord string, endWord string, wordList []string) int {
	set := make(map[string]int)
	for i := 0; i < len(wordList); i++ {
		set[wordList[i]] = 1
	}
	if _, ok := set[endWord]; !ok {
		return 0
	}

	m := make(map[string]int)
	m[beginWord] = 1

	q := []string{beginWord}

	for len(q) > 0 {
		word := q[0]
		q = q[1:]
		for i := 0; i < len(word); i++ {
			byteWord := []byte(word)
			for c := 'a'; c <= 'z'; c++ {
				byteWord[i] = byte(c)
				newWord := string(byteWord)
				if set[newWord] > 0 && newWord == endWord {
					return m[word] + 1
				}
				if set[newWord] > 0 && m[newWord] == 0 {
					m[newWord] = m[word] + 1
					q = append(q, newWord)
				}
			}
		}
	}
	return 0
}
